### PR TITLE
Support for setting the date/time

### PR DESCRIPTION
**Pull request by Willem3141**
This adds the option to set the date/time in the settings app.

I'm not sure if it is desirable to merge this now already, since the kernel function to actually set the time (`convertTimeToTicks`) is not done. I'm creating this pull request now anyway to get some comments, and to have a place to discuss this feature.

### CODE REVIEW

File: datetime.asm [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+unsupported:
+    
+    kld(hl, clock_unsupported_message)
+    kld(de, clock_unsupported_options)
+    ld a, 0
+    ld b, 0
+    corelib(showMessage)
+    
+    ret
```
Comment (SirCmpwn) [inline datetime.asm]: I think it would be better to simply omit the clock option from the earlier screen if the clock is unsupported.

**Comment by SirCmpwn** [comment]
First pass over the code looks good. Having a look at functionality now.

**Comment by SirCmpwn** [comment]
Functionality looks awesome. This is super cool.

One change I would make is to center the date/time interface on the screen, rather than aligning it to the left.

**Comment by SirCmpwn** [comment]
FYI, I added some features to settings today and you'll have to merge from upstream before this can be applied. Am I correct that this can be merged now and will Just Work when the relevant feature is added to the kernel?

**Comment by Willem3141** [comment]
Yes, it should start working when the kernel function is added.

At the moment there is one remaining issue, namely that the date wrap-around code assumes all months have 31 days. (So if you decrease the day when 2015-03-01 is selected, you get 2015-02-31 instead of 2015-02-28.) I'll hopefully have time to look into that today.

**Comment by SirCmpwn** [comment]
Cool. I'm going to start doing a weekly mailing list blast of "this week in KnightOS", and it'd be cool if this could be included in the first one on friday :wink: 

It also looks like leap years are not handled - when you tidy up all those things, let me know and I'll give this a merge (without kernel support).

**Comment by Willem3141** [comment]
I'm working on the leap years right now :)

**Comment by SirCmpwn** [comment]
Great!

**Comment by SirCmpwn** [comment]
Head's up - you need to handle losing focus correctly (by redrawing the screen when you regain focus). You'll know that you've losed focus because `corelib(app{Get,Wait}Key)` will reset Z.

**Comment by Willem3141** [comment]
I think all the outstanding issues are resolved now, except for the suggestion to omit the clock option from the menu if the calculator doesn't have a clock. I agree that that would be nicer from a user perspective, and I can build that, but it would complicate the menu code in `main.asm` quite a lot, I think... Do you think that is worth it?

**Comment by EphraimB** [comment]
Yes

**Comment by SirCmpwn** [comment]
Yes, I think it's worth it. I would rather not merge this without. It should be just as simple as `pcall(getClock)`, then load a different menu table if NZ.

**Comment by SirCmpwn** [comment]
We should have a `clockSupported` kernel call, by the way.

**Comment by Willem3141** [comment]
I have that kernel function on my list of useful kernel functions that I want to add :smiley:

**Comment by EphraimB** [comment]
How many more stuff needs to be done until ConvertTimeToTicks is finished?

**Comment by SirCmpwn** [comment]
Date/time support in general is being tracked here: https://github.com/KnightOS/KnightOS/issues/168

**Comment by Willem3141** [comment]
I think it is ready; the menu item is not shown anymore when the clock is unsupported :smile:

**Comment by SirCmpwn** [comment]
Awesome!

**Pull request merged**